packages/react-reconciler/src/ReactFiberHooks.js

```diff
@@ -1232,9 +1232,14 @@ function useMemoCache(size: number): Array<any> {
       data[i] = REACT_MEMO_CACHE_SENTINEL;
     }
   } else if (data.length !== size) {
-    // TODO: consider warning or throwing here
     if (__DEV__) {
-      console.error(
+      // reset cache for FastRefresh
+      data = memoCache.data[memoCache.index] = new Array(size);
+      for (let i = 0; i < size; i++) {
+        data[i] = REACT_MEMO_CACHE_SENTINEL;
+      }
```

Comment:
An alternative approach might be to append a debug property on the fiber during fast refresh to reset the cache only when Fresh actually kicks in. That has the advantage of resetting the cache even when cache size changes

Reply:
This seems to be the better approach to me.

- React shouldn't be throwing prod-only errors.
- Feels very likely there are types of changes where the memo size doesn't change (e.g. some literal such as a style is updated in the component)

Reply:
What does useMemo do if its implementation changes? Does it just keep its cache entry? Seems like it would need to reset for the same reason.

Reply:
useMemo always drops the cache on hot reload. Seems like we should do the same for uMC? We assume memoization is not essential for the logic and can be reconstructed. This also tests that the component logic is resilient to it somewhat.

Reply:
https://github.com/facebook/react/pull/30700